File: servers/typescript/src/data_store.ts
export type DataEntry = {
  data: unknown;
  version: number;
  updatedBy: string;
};

export class ConflictError extends Error {
  readonly actualVersion: number;
  readonly currentData: unknown;

  constructor(actualVersion: number, currentData: unknown) {
    super(`version conflict: actual version is ${actualVersion}`);
    this.actualVersion = actualVersion;
    this.currentData = currentData;
  }
}

const EMPTY_ENTRY: DataEntry = { data: undefined, version: 0, updatedBy: "" };

export class DataStore {
  private session = new Map<string, DataEntry>();
  private client = new Map<string, Map<string, DataEntry>>();

  apply(
    op: string,
    key: string,
    scope: string,
    clientId: string,
    data: unknown,
    expectedVersion?: number,
  ): DataEntry {
    const store = this.storeForScope(scope, clientId);
    const existing = store.get(key) ?? EMPTY_ENTRY;

    if (expectedVersion !== undefined && existing.version !== expectedVersion) {
      throw new ConflictError(existing.version, existing.data);
    }

    if (op === "delete") {
      store.delete(key);
      return { data: undefined, version: existing.version + 1, updatedBy: clientId };
    }

    const newData = applyOp(op, existing.data, data);
    const entry: DataEntry = {
      data: newData,
      version: existing.version + 1,
      updatedBy: clientId,
    };
    store.set(key, entry);
    return entry;
  }

  get(key: string, scope: string, clientId: string): DataEntry {
    const store = this.storeForScope(scope, clientId);
    return store.get(key) ?? EMPTY_ENTRY;
  }

  private storeForScope(scope: string, clientId: string): Map<string, DataEntry> {
    if (scope === "self") {
      let clientStore = this.client.get(clientId);
      if (!clientStore) {
        clientStore = new Map();
        this.client.set(clientId, clientStore);
      }
      return clientStore;
    }
    return this.session;
  }
}

function applyOp(op: string, existing: unknown, incoming: unknown): unknown {
  switch (op) {
    case "replace":
      return incoming;
    case "merge":
      return mergeObjects(existing, incoming);
    case "set.add":
      return setAdd(existing, incoming);
    case "set.remove":
      return setRemove(existing, incoming);
    case "list.add":
      return listAdd(existing, incoming);
    case "list.remove":
      return listRemove(existing, incoming);
    case "counter.add":
      return counterAdd(existing, incoming);
    default:
      throw new Error(`invalid operation: ${op}`);
  }
}

function mergeObjects(existing: unknown, incoming: unknown): unknown {
  if (typeof incoming !== "object" || incoming === null || Array.isArray(incoming)) {
    throw new Error("merge requires an object");
  }
  const base: Record<string, unknown> =
    typeof existing === "object" && existing !== null && !Array.isArray(existing)
      ? { ...(existing as Record<string, unknown>) }
      : {};
  return { ...base, ...(incoming as Record<string, unknown>) };
}

function setAdd(existing: unknown, incoming: unknown): unknown[] {
  const set: unknown[] = Array.isArray(existing) ? [...existing] : [];
  const inStr = JSON.stringify(incoming);
  for (const item of set) {
    if (JSON.stringify(item) === inStr) return set;
  }
  set.push(incoming);
  return set;
}

function setRemove(existing: unknown, incoming: unknown): unknown[] {
  const set: unknown[] = Array.isArray(existing) ? existing : [];
  const inStr = JSON.stringify(incoming);
  return set.filter((item) => JSON.stringify(item) !== inStr);
}

function listAdd(existing: unknown, incoming: unknown): unknown[] {
  const list: unknown[] = Array.isArray(existing) ? [...existing] : [];
  list.push(incoming);
  return list;
}

function listRemove(existing: unknown, incoming: unknown): unknown[] {
  const list: unknown[] = Array.isArray(existing) ? existing : [];
  const inStr = JSON.stringify(incoming);
  return list.filter((item) => JSON.stringify(item) !== inStr);
}

function counterAdd(existing: unknown, incoming: unknown): number {
  const current = typeof existing === "number" ? existing : 0;
  if (typeof incoming !== "number") {
    throw new Error("counter.add requires a number");
  }
  return current + incoming;
}

File: servers/python/src/starfish_server/data_store.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any


@dataclass
class DataEntry:
    data: Any
    version: int
    updated_by: str


class ConflictError(Exception):
    def __init__(self, actual_version: int, current_data: Any) -> None:
        super().__init__(f"version conflict: actual version is {actual_version}")
        self.actual_version = actual_version
        self.current_data = current_data


EMPTY_ENTRY = DataEntry(data=None, version=0, updated_by="")


class DataStore:
    def __init__(self) -> None:
        self._session: dict[str, DataEntry] = {}
        self._client: dict[str, dict[str, DataEntry]] = {}

    def apply(
        self,
        op: str,
        key: str,
        scope: str,
        client_id: str,
        data: Any,
        expected_version: int | None = None,
    ) -> DataEntry:
        store = self._store_for_scope(scope, client_id)
        existing = store.get(key, EMPTY_ENTRY)

        if expected_version is not None and existing.version != expected_version:
            raise ConflictError(existing.version, existing.data)

        if op == "delete":
            store.pop(key, None)
            return DataEntry(data=None, version=existing.version + 1, updated_by=client_id)

        new_data = _apply_op(op, existing.data, data)
        entry = DataEntry(data=new_data, version=existing.version + 1, updated_by=client_id)
        store[key] = entry
        return entry

    def get(self, key: str, scope: str, client_id: str) -> DataEntry:
        store = self._store_for_scope(scope, client_id)
        return store.get(key, EMPTY_ENTRY)

    def _store_for_scope(self, scope: str, client_id: str) -> dict[str, DataEntry]:
        if scope == "self":
            if client_id not in self._client:
                self._client[client_id] = {}
            return self._client[client_id]
        return self._session


def _apply_op(op: str, existing: Any, incoming: Any) -> Any:
    match op:
        case "replace":
            return incoming
        case "merge":
            return _merge_objects(existing, incoming)
        case "set.add":
            return _set_add(existing, incoming)
        case "set.remove":
            return _set_remove(existing, incoming)
        case "list.add":
            return _list_add(existing, incoming)
        case "list.remove":
            return _list_remove(existing, incoming)
        case "counter.add":
            return _counter_add(existing, incoming)
        case _:
            raise ValueError(f"invalid operation: {op}")


def _merge_objects(existing: Any, incoming: Any) -> Any:
    if not isinstance(incoming, dict):
        raise ValueError("merge requires an object")
    base: dict[str, Any] = {}
    if isinstance(existing, dict):
        base = {**existing}
    return {**base, **incoming}


def _set_add(existing: Any, incoming: Any) -> list[Any]:
    items: list[Any] = list(existing) if isinstance(existing, list) else []
    in_str = json.dumps(incoming, sort_keys=True)
    for item in items:
        if json.dumps(item, sort_keys=True) == in_str:
            return items
    items.append(incoming)
    return items


def _set_remove(existing: Any, incoming: Any) -> list[Any]:
    items: list[Any] = existing if isinstance(existing, list) else []
    in_str = json.dumps(incoming, sort_keys=True)
    return [item for item in items if json.dumps(item, sort_keys=True) != in_str]


def _list_add(existing: Any, incoming: Any) -> list[Any]:
    items: list[Any] = list(existing) if isinstance(existing, list) else []
    items.append(incoming)
    return items


def _list_remove(existing: Any, incoming: Any) -> list[Any]:
    items: list[Any] = existing if isinstance(existing, list) else []
    in_str = json.dumps(incoming, sort_keys=True)
    return [item for item in items if json.dumps(item, sort_keys=True) != in_str]


def _counter_add(existing: Any, incoming: Any) -> int | float:
    current = existing if isinstance(existing, (int, float)) else 0
    if not isinstance(incoming, (int, float)):
        raise ValueError("counter.add requires a number")
    return current + incoming
